client/devices/client.go

```diff
@@ -57,81 +59,109 @@ type Client struct {
 	url            string
 	devicesListURL string
 	client         *http.Client
+	output         io.Writer
 }
 
 func NewClient(url string, skipVerify bool) *Client {
 	return &Client{
 		url:            url,
 		devicesListURL: client.JoinURL(url, devicesListURL),
 		client:         client.NewHttpClient(skipVerify),
+		output:         os.Stdout,
 	}
 }
 
-func (c *Client) ListDevices(token string, detailLevel int, raw bool) error {
+func (c *Client) ListDevices(token string, detailLevel, perPage, page int, raw bool) error {
 	if detailLevel > 3 || detailLevel < 0 {
 		return fmt.Errorf("FAILURE: invalid devices detail")
 	}
 
-	body, err := client.DoGetRequest(token, c.devicesListURL, c.client)
+	req, err := http.NewRequest(http.MethodGet, c.devicesListURL, nil)
+	if err != nil {
+		return fmt.Errorf("failed to prepare request: %w", err)
+	}
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
+	q := url.Values{
+		"per_page": []string{strconv.Itoa(perPage)},
+		"page":     []string{strconv.Itoa(page)},
+	}
+	req.URL.RawQuery = q.Encode()
+
+	reqDump, err := httputil.DumpRequest(req, false)
+	if err != nil {
+		return err
+	}
+	log.Verbf("sending request: \n%s", string(reqDump))
+
+	rsp, err := c.client.Do(req)
 	if err != nil {
 		return err
 	}
+	defer rsp.Body.Close()
+	if rsp.StatusCode != 200 {
+		return fmt.Errorf("GET %s request failed with status %d",
+			req.URL.RequestURI(), rsp.StatusCode)
+	}
 
 	if raw {
-		fmt.Println(string(body))
+		_, err := io.Copy(os.Stdout, rsp.Body)
```

Comment:
this will not be pretty formatted, right?

Reply:
No formatting, just the raw (JSON) response.